lib/ABI/FunctionType/Conversion.cpp

```diff
@@ -458,41 +464,116 @@ TCC::tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
 
-    // TODO: ensure that the type is in fact naturally aligned
-    New.Type() = CurrentArgument.Type();
+    CurrentRange = std::ranges::subrange(std::next(CurrentRange.begin()),
+                                         Stack.Fields().end());
   }
 
-  // And don't forget to convert the very last field.
-  const model::StructField &LastArgument = *std::prev(Stack.Fields().end());
-  model::Argument &New = Result.emplace_back();
-  model::copyMetadata(New, LastArgument);
-  New.Index() = IndexOffset++;
+  // The main loop is over. Which means that we either converted everything but
+  // the very last argument correctly, OR that we aborted half-way through.
+  if (CurrentRange.size() == 1) {
+    // Having only one element in the "remaining" range means that only the last
+    // field is left - add it too after checking.
+    const model::StructField &LastArgument = CurrentRange.front();
+
+    // This is a workaround for the cases where expected alignment is missing
+    // at the very end of the RFT-style stack argument struct:
+    uint64_t FullSize = abi::FunctionType::paddedSizeOnStack(Stack.Size(),
+                                                             FullAlignment);
+
+    std::optional<uint64_t> LastSize = LastArgument.Type().size();
+    revng_assert(LastSize.has_value() && LastSize.value() != 0);
+    if (canAddArgumentAsIs(LastArgument.Type(),
+                           LastArgument.Offset(),
+                           FullSize,
+                           FullAlignment,
+                           Result.size(),
+                           UsedGPRegisterCount,
+                           UsedVRegisterCount,
+                           CurrentStackOffset)) {
+      model::Argument &New = Result.emplace_back();
+      model::copyMetadata(New, LastArgument);
+      New.Type() = LastArgument.Type();
+      New.Index() = IndexOffset++;
+      return Result;
+    }
+  }
 
-  // TODO: ensure that the type is in fact naturally aligned
-  New.Type() = LastArgument.Type();
+  // Getting to this point (past the return statement in the last element
+  // section) means that there is at least one argument we cannot just "add".
+  revng_log(Log,
+            "Naive conversion failed. Try to fall back on using structs "
+            "instead.");
+  if (CurrentRange.size() != Stack.Fields().size()) {
+    // This condition being true means that we did succeed in converting some
+    // of the arguments, but failed on some others. Let's try to wrap
+    // the remainder into a struct and see if bundled together they make more
+    // sense in c-like representation.
+    revng_log(Log,
+              "Some fields were converted successfully, try to slot in the "
+              "rest as a struct.");
+    const model::StructField &LastSuccessful = *std::prev(CurrentRange.begin());
+    std::uint64_t Offset = LastSuccessful.Offset();
+    Offset += ABI.paddedSizeOnStack(*LastSuccessful.Type().size());
+
+    model::StructType> ResultType;
```

Comment:
I don't think this compiles

Reply:
Yeah, I noticed it right after pushing and repushed fix right away :D